lib/FunctionIsolation/PromoteCSVs.cpp

```diff
@@ -572,26 +576,40 @@ void PromoteCSVs::wrapCallsToHelpers(Function *F) {
   }
 }
 
-void PromoteCSVs::run() {
-  for (Function &F : FunctionTags::ABIEnforced.functions(M)) {
-    // Add tag
-    FunctionTags::CSVsPromoted.addTo(&F);
+void PromoteCSVs::runOnFunction(llvm::Function &F) {
+  // Add tag
+  FunctionTags::CSVsPromoted.addTo(&F);
 
-    if (not F.isDeclaration()) {
-      // Wrap calls to wrappers
-      wrapCallsToHelpers(&F);
+  if (not F.isDeclaration()) {
+    // Wrap calls to wrappers
+    wrapCallsToHelpers(&F);
 
-      // (Re-)promote CSVs
-      promoteCSVs(&F);
-    }
+    // (Re-)promote CSVs
+    promoteCSVs(&F);
+  }
+}
+
+void PromoteCSVs::run() {
+  for (Function &F : FunctionTags::ABIEnforced.functions(M)) {
+    runOnFunction(F);
   }
 }
 
-bool PromoteCSVsPass::runOnModule(Module &M) {
+bool PromoteCSVsPassImpl::prologue(llvm::Module &Module,
+                                   const model::Binary &Binary) {
+
   auto &GCBI = getAnalysis<GeneratedCodeBasicInfoWrapperPass>().getGCBI();
   auto &ModelWrapper = getAnalysis<LoadModelWrapperPass>().get();
-  const model::Binary &Binary = *ModelWrapper.getReadOnlyModel();
-  PromoteCSVs HW(&M, GCBI, Binary);
-  HW.run();
+  HW = new PromoteCSVs(&Module, GCBI, Binary);
```

Comment:
Absolutely not.

Reply:
i can remove the new invocation here, but it would imply that PromoteCSVs is moved to a header